### PR TITLE
test(helm): Run smoke test in multi-node cluster

### PR DESCRIPTION
**Pull request by sayboras**
To run the smoke test in kind with multi nodes, the steps are mentioned
in https://docs.cilium.io/en/latest/gettingstarted/kind/

Remove chart-testing as it's a little bit restricted.

Signed-off-by: Tam Mach <sayboras@yahoo.com>

```release-note
test(helm): Run smoke test in multi-node cluster
```

### CODE REVIEW

File: .github/kind-config.yaml
Comment (nebril) [inline .github/kind-config.yaml]: I think we only need 2 nodes for conformance tests to run, let's remove 2 of these workers, this should speed up cluster creation (this is supposed to be a really fast check).

**Review by nebril — CHANGES_REQUESTED** [review]
Great improvements! One change request left inline, after it's fixed I think we should merge it. Thanks!

File: .github/kind-config.yaml
Comment (sayboras) [reply .github/kind-config.yaml]: Good point 💯, let me remove two. Thanks

Comment (sayboras) [reply .github/kind-config.yaml]: seems like we will require 2 worker nodes for connectivity check.

Here is one run with only 1 worker node https://github.com/cilium/cilium/runs/740518981?check_suite_focus=true. This run was failed.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
@nebril do you want this to be backported to 1.8?

**Comment by errordeveloper** [comment]
@aanm is the kind test already in 1.8 branch and are we blocking on it?

From my point it's not clear if kind is reliable enough to be blocking PR on, we haven't fully assessed how datapath works in multi-node secenrio specifically. My previous experiment of running kind-like multi-node setup resulted in mixed results, where in some case nodes had completely broken networking.

I also expect usage of kind to evolve in master during the 1.9 cycle, so I'd be agains having an early version in 1.8 branch or back-porting every improvement. On related topic, I also would like to find a way of minimising CI infra back-ports.

File: .github/kind-config.yaml
Comment (nebril) [reply .github/kind-config.yaml]: @sayboras  That's because kind-control-plane node is tainted as master. This can be fixed by running `kubectl taint node kind-control-plane node-role.kubernetes.io/master-`, as we don't really care about pods crashing our master node.

Comment (sayboras) [reply .github/kind-config.yaml]: @nebril I have added the configuration to make sure there is no taint for control plane node :). Thanks for your input.

File: .github/workflows/helm-check.yaml
Comment (pchaigno) [inline .github/workflows/helm-check.yaml]: Any reason the `K8sConformance` test isn't executed anymore?

**Review by pchaigno — CHANGES_REQUESTED** [review]
We are planning to make the smoke test + Travis CI build required before launching e2e tests. The goal is to catch common mistakes before spinning up several more expensive VMs.

We therefore want to keep these two first tests as quick as possible. One concern that was raised during Cilium's community meeting is that running the smoke test in a multi-node cluster might slow it down. It isn't really a concern if it remains faster than the Travis CI build (~21 minutes if we don't consider aarch64 build as required), but it's something we'll need to look into.

It looks like the `K8sConformance` test isn't actually executed currently (comment :point_down: ) so it's hard to say if the smoke test remains faster to Travis CI build.

File: .github/workflows/helm-check.yaml
Comment (sayboras) [reply .github/workflows/helm-check.yaml]: I think the existing conformance test is to perform connectivity check for _single node_ . This changes are to extend it with multiple nodes, some syntactic sugar with helm + kubectl apply to reduce dependency download time.

Based on the recent run, it is taking 10 mins (5 mins for docker build, and 3 mins for cilium installation + connectivity check).

Let me know what you have in mind :).

Comment (pchaigno) [reply .github/workflows/helm-check.yaml]: You're right, it's running https://github.com/cilium/cilium/blob/819185568eb5cf653052043c5299d1224460368b/test/k8sT/Conformance.go#L83
I think you want to run this instead: https://github.com/cilium/cilium/blob/819185568eb5cf653052043c5299d1224460368b/test/k8sT/Conformance.go#L69

So the following should do it:
```diff
-           ginkgo -v --focus="K8sConformance Portmap Chaining Check one node" -- -cilium.provision=false -test.v --cilium.kubeconfig=${HOME}/.kube/config -cilium.image=cilium/cilium:latest -cilium.operator-image=cilium/operator-generic:latest
+           ginkgo -v --focus="K8sConformance Portmap Chaining Check connectivity-check" -- -cilium.provision=false -test.v --cilium.kubeconfig=${HOME}/.kube/config -cilium.image=cilium/cilium:latest -cilium.operator-image=cilium/operator-generic:latest
```

Comment (pchaigno) [reply .github/workflows/helm-check.yaml]: Discussed offline. I misunderstood this initially.
It doesn't run ginkgo on purpose and just does a simple connectivity check. Given that the smoke test now takes 10min where it was taking 15min before, I'm :+1: with this if @cilium/ci folks are.

**Review by pchaigno — APPROVED** [review]
Smoke test takes ~10min to run with this PR, so :+1: for me.

**Comment by sayboras** [comment]
Sorry for so many forced pushes. Some of them for the rebase with master, and few are for debugging the flanky test due to CRD not available, which is fixed by `kubectl wait crd` command.

**Review by tgraf — APPROVED** [review]

File: .github/workflows/smoke-test.yaml
Comment (rolinh) [inline .github/workflows/smoke-test.yaml:82]: I think we want to add `--set config.ipam=kubernetes` here so that cilium picks up the range defined by kind that we may adjust via the kind config file if we have a network conflict (see #12181).

Comment (sayboras) [reply .github/workflows/smoke-test.yaml:82]: Thanks for your input, I have added the same :)

**Comment by pchaigno** [comment]
test-me-please

**Review by rolinh — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Pull request merged**
